pkgs/http/lib/src/base_response.dart

```diff
@@ -68,3 +71,34 @@ abstract class BaseResponse {
     }
   }
 }
+
+/// Fields and methods that will be added to [BaseResponse] when `package:http`
+/// version 2 is released.
+///
+/// [Client] methods that return a [BaseResponse] subclass, such as [Response]
+/// or [StreamedResponse], **may** return a [BaseResponseV2].
+///
+/// For example:
+///
+/// ```dart
+/// final client = Client();
+/// final response = client.get(Uri.https('example.com', '/'));
+/// Uri? finalUri;
+/// if (response is BaseResponseV2) {
+///   finalUri = (response as BaseResponseV2).uri;
+/// }
+/// // Do something with `finalUri`.
+/// client.close();
+/// ```
+mixin BaseResponseV2 on BaseResponse {
+  /// The [Uri] of the response returned by the server.
+  ///
+  /// If no redirects were followed, [url] will be the same as the requested
+  /// [Uri].
+  ///
+  /// If redirects were followed, [url] will be the [Uri] of the last redirect
+  /// that was followed.
+  ///
+  /// May be `null` when the response is for a special URL such as "about:".
```

Comment:
Double checking - is this a limitation of the Web platform?

Would we be able to make this non-nullable if we override a `null` one from the browser back to the request Uri?

Reply:
Agreed. What is the purpose of handling URLs such as "about:"? IMO they cannot exist in responses.

Reply:
The standards docs for `XMLHttpRequest` aren't clear on when this can happen, but [`XMLHttpRequest.responseUrl`](https://xhr.spec.whatwg.org/#the-responseurl-attribute) can be the empty string. 

I'll make it non-nullable we can hope for the best.